File: lua/difference-of-squares/difference-of-squares.lua
local square_of_sum = (function()
    local function square_of_sum(n, squares_sum)
        if n > 0 then
            return square_of_sum(n - 1, squares_sum + n)
        else
            return squares_sum ^ 2
        end
    end

    return function(n)
        return square_of_sum(n, 0)
    end
end)()


local sum_of_squares = (function()
    local function sum_of_squares(n, sum_squares)
        if n > 0 then
            return sum_of_squares(n - 1, sum_squares + n ^ 2)
        else
            return sum_squares
        end
    end

    return function(n)
        return sum_of_squares(n, 0)
    end
end)()

local function difference_of_squares(n)
    return square_of_sum(n) - sum_of_squares(n)
end

return {
    square_of_sum = square_of_sum,
    sum_of_squares = sum_of_squares,
    difference_of_squares = difference_of_squares
}

File: cpp/difference-of-squares/difference_of_squares.cpp
#include "difference_of_squares.h"
#include <cmath>

namespace difference_of_squares {

int do_sum_of_squares(int current, int previous_sum) {
    if(current < 0) {
        return -1;
    }
    if(current == 0) {
        return previous_sum;
    }
    
    const int square = std::pow(current, 2);
    const int next = current - 1;

    return do_sum_of_squares(next, previous_sum + square);
}

int do_square_of_sum(int current, int previous_sum) {
    if(current < 0) {
        return -1;
    }
    if(current == 0) {
        return std::pow(previous_sum, 2);
    }
    const int next = current - 1;
    const int current_sum = previous_sum + current;
    return do_square_of_sum(next, current_sum);
}

int sum_of_squares(int last) {
    return do_sum_of_squares(last, 0);
}

int square_of_sum(int last) {
    return do_square_of_sum(last, 0);
}

int difference(int last) {
    return square_of_sum(last) - sum_of_squares(last);  
}

}  // namespace difference_of_squares
